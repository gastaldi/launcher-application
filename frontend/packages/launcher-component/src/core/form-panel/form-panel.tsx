import * as React from 'react';
import { useState } from 'react';
import { Button, Toolbar, ToolbarGroup } from '@patternfly/react-core';
import { InputProps } from '../types';
import { Separator } from '../stuff';

interface FormPanelProps<T> {
  id?: string;
  initialValue: T;

  validator?(value: T): boolean;

  children(inputProps: InputProps<T>): any;

  onSave?(value: T);

  onCancel?();
}

export function FormPanel<T>(props: FormPanelProps<T>) {
  const [value, onChange] = useState<T>(props.initialValue);

  const onSave = () => {
    if (props.onSave) {
      props.onSave(value);
    }
  };

  const onCancel = () => {
    onChange(props.initialValue);
    if (props.onCancel) {
      props.onCancel();
    }
  };
  const id = props.id || 'form';
  return (
    <div className="form-panel" style={{padding: '20px'}}>
      {props.children({value, onChange})}
      <Separator/>
      <Toolbar>
        <ToolbarGroup>
          <Button
            variant="primary"
            aria-label={'Save ' + id}
            onClick={onSave}
            isDisabled={props.validator && !props.validator(value)}
          >
            Save
          </Button>
        </ToolbarGroup>
        <ToolbarGroup>
          <Button variant="secondary" aria-label={'Cancel ' + id} onClick={onCancel}>Cancel</Button>
        </ToolbarGroup>
      </Toolbar>
    </div>
  );
}
