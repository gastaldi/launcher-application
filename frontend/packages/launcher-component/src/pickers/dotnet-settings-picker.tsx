import { Form } from '@patternfly/react-core';
import React from 'react';
import { InputProps, Picker } from '../core/types';
import { LaunchTextInput } from '../core/text-input/text-input';

export interface DotNetSettingsPickerValue {
  namespace?: string;
  version?: string;
}

interface DotNetSettingsPickerProps extends InputProps<DotNetSettingsPickerValue> {
}

// tslint:disable-next-line:max-line-length
const VERSION_REGEXP = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(-(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(\.(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*)?(\+[0-9a-zA-Z-]+(\.[0-9a-zA-Z-]+)*)?$/;
const NAME_REGEXP = /^[a-zA-Z_].*$/;

export const DotNetSettingsPicker: Picker<DotNetSettingsPickerProps, DotNetSettingsPickerValue> = {
  checkCompletion: value => !!value.namespace && NAME_REGEXP.test(value.namespace)
    && !!value.version && VERSION_REGEXP.test(value.version),
  Element: props => {
    return (
      <Form>
        <LaunchTextInput
          label="Namespace"
          isRequired
          helperTextInvalid="Please provide a valid namespace"
          type="text"
          id="name"
          name="name"
          aria-label=".Net namespace"
          value={props.value.namespace || ''}
          onChange={value => props.onChange({ ...props.value, namespace: value })}
          pattern={NAME_REGEXP.source}
          isValid={NAME_REGEXP.test(props.value.namespace || '')}
        />
        <LaunchTextInput
          label="Version"
          isRequired
          helperTextInvalid="Please provide a valid version number"
          type="text"
          id="version"
          name="version"
          aria-label=".Net version"
          value={props.value.version || ''}
          onChange={value => props.onChange({ ...props.value, version: value })}
          isValid={VERSION_REGEXP.test(props.value.version || '')}
        />
      </Form >
    );
  }
};
