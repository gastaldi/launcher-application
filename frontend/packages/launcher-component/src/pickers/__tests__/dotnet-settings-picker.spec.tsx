import React from 'react';
import { render, fireEvent } from 'react-testing-library';

import { FormPanel } from '../../core/form-panel/form-panel';
import { DotNetSettingsPicker } from '../dotnet-settings-picker';

describe('<DotNetSettingsPicker />', () => {
  it('renders the DotNetSettingsPicker correctly', () => {
    const comp = render(<DotNetSettingsPicker.Element value={{name: '', version: ''}} onChange={() => {}}/>);
    expect(comp.asFragment()).toMatchSnapshot();
  });

  it('show error for invalid data', () => {
    const handleSave = jest.fn();
    const comp = render(
      <FormPanel
        initialValue={{}}
        validator={DotNetSettingsPicker.checkCompletion}
        onSave={handleSave}
        onCancel={() => {}}
      >
        {(inputProps) => (<DotNetSettingsPicker.Element {...inputProps}/>)}
      </FormPanel>
    );

    const nameField = comp.getByLabelText('.Net namespace');
    fireEvent.change(nameField, { target: { value: 'invalid name' } });
    const versionField = comp.getByLabelText('.Net version');
    fireEvent.change(versionField, { target: { value: '1' } });
    expect(comp.asFragment()).toMatchSnapshot();
  });
});
