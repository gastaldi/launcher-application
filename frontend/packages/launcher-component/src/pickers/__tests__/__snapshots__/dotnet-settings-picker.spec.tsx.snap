// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DotNetSettingsPicker /> renders the DotNetSettingsPicker correctly 1`] = `
<DocumentFragment>
  <form
    class="pf-c-form"
    novalidate=""
  >
    <div
      class="pf-c-form__group"
    >
      <label
        class="pf-c-form__label"
        for="name"
      >
        Namespace
      </label>
      <input
        aria-invalid="false"
        aria-label=".Net namespace"
        class="pf-c-form-control"
        id="name"
        name="name"
        pattern="^[a-zA-Z_].*$"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="pf-c-form__group"
    >
      <label
        class="pf-c-form__label"
        for="version"
      >
        Version
      </label>
      <input
        aria-invalid="false"
        aria-label=".Net version"
        class="pf-c-form-control"
        id="version"
        name="version"
        required=""
        type="text"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`<DotNetSettingsPicker /> show error for invalid data 1`] = `
<DocumentFragment>
  <div
    class="form-panel"
    style="padding: 20px;"
  >
    <form
      class="pf-c-form"
      novalidate=""
    >
      <div
        class="pf-c-form__group"
      >
        <label
          class="pf-c-form__label"
          for="name"
        >
          Namespace
        </label>
        <input
          aria-invalid="false"
          aria-label=".Net namespace"
          class="pf-c-form-control"
          id="name"
          name="name"
          pattern="^[a-zA-Z_].*$"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="pf-c-form__group"
      >
        <label
          class="pf-c-form__label"
          for="version"
        >
          Version
        </label>
        <input
          aria-invalid="false"
          aria-label=".Net version"
          class="pf-c-form-control"
          id="version"
          name="version"
          required=""
          type="text"
          value=""
        />
      </div>
    </form>
    <hr
      class="separator"
    />
    <div
      class="pf-l-toolbar"
    >
      <div
        class="pf-l-toolbar__group"
      >
        <button
          aria-label="Save form"
          class="pf-c-button pf-m-primary pf-m-disabled"
          disabled=""
          type="button"
        >
          Save
        </button>
      </div>
      <div
        class="pf-l-toolbar__group"
      >
        <button
          aria-label="Cancel form"
          class="pf-c-button pf-m-secondary"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
